src/Lean/Data/PersistentHashMap.lean

```diff
@@ -70,20 +70,24 @@ private theorem size_set {ks : Array α} {vs : Array β} (h : ks.size = vs.size)
 private theorem size_push {ks : Array α} {vs : Array β} (h : ks.size = vs.size) (k : α) (v : β) : (ks.push k).size = (vs.push v).size := by
   simp [h]
 
-partial def insertAtCollisionNodeAux [BEq α] : CollisionNode α β → Nat → α → β → CollisionNode α β
+instance : Inhabited (CollisionNode α β × Bool) where
+  default := (⟨Node.collision #[] #[] rfl, .mk _ _ _⟩, true)
+
+partial def insertAtCollisionNodeAux [BEq α] : CollisionNode α β → Nat → α → β → CollisionNode α β × Bool
   | n@⟨Node.collision keys vals heq, _⟩, i, k, v =>
     if h : i < keys.size then
       let idx : Fin keys.size := ⟨i, h⟩;
       let k' := keys.get idx;
       if k == k' then
          let j : Fin vals.size := ⟨i, by rw [←heq]; assumption⟩
-         ⟨Node.collision (keys.set idx k) (vals.set j v) (size_set heq idx j k v), IsCollisionNode.mk _ _ _⟩
+         (⟨Node.collision (keys.set idx k) (vals.set j v) (size_set heq idx j k v), IsCollisionNode.mk _ _ _⟩, true)
       else insertAtCollisionNodeAux n (i+1) k v
     else
-      ⟨Node.collision (keys.push k) (vals.push v) (size_push heq k v), IsCollisionNode.mk _ _ _⟩
+      (⟨Node.collision (keys.push k) (vals.push v) (size_push heq k v), IsCollisionNode.mk _ _ _⟩, false)
   | ⟨Node.entries _, h⟩, _, _, _ => False.elim (nomatch h)
 
-def insertAtCollisionNode [BEq α] : CollisionNode α β → α → β → CollisionNode α β :=
+/-- Inserts a key-value pair into a CollisionNode, also returning whether an existing value was replaced. -/
+def insertAtCollisionNode [BEq α] : CollisionNode α β → α → β → CollisionNode α β × Bool  :=
   fun n k v => insertAtCollisionNodeAux n 0 k v
```

Comment:
If the overhead from the pair is contributing to performance loss, then I think it would be to make `insertAtCollisionNode` not recursive and then inline it into `insertAux` for further optimization.  A non-recursive definition may be something like the following:
```lean
def insertAtCollisionNode [BEq α] : CollisionNode α β → α → β → CollisionNode α β × Bool  :=
| ⟨Node.collision keys vals heq, _⟩, k, v =>
    let i := (keys.findIdx? k).getD key.size 
    if h : i < keys.size then
      let idx : Fin keys.size := ⟨i, h⟩;
      let j : Fin vals.size := ⟨i, by rw [←heq]; assumption⟩
      (⟨Node.collision (keys.set idx k) (vals.set j v) (size_set heq idx j k v), IsCollisionNode.mk _ _ _⟩, true)
    else
      (⟨Node.collision (keys.push k) (vals.push v) (size_push heq k v), IsCollisionNode.mk _ _ _⟩, false)
| ⟨Node.entries _, h⟩, _, _ => False.elim (nomatch h)
```

Reply:
Looks like the performance loss was not real, so I'm not going to pursue this.